### DISCUSSION TITLE
Selectively removing interactions from molecule

### DISCUSSION THREAD

**Discussion by csbrasnett** [marrink-lab/vermouth-martinize#482] [General]
I need to edit an itp to remove parts of the elastic network, as determined by secondary structure assignment.

As a minimal example, I would have thought something like this should work:

```python
# loop over regions where we want to keep EN
for i in wanted_region:
     # look at all the elastic network interactions
     for j in all_EN_interactions:
         # compare atom indicies of the EN interaction to the boundaries of the region we want to keep
         # if they're within, keep them
         if all_EN_interactions[j] in wanted_region[i]:
             continue
         # if they're outside, remove the interaction
         else:
             ff.blocks[molname].remove_interaction('bonds', atoms = all_EN_interactions[j])
```

However, I keep getting a KeyError
`KeyError: "Can't find interaction of type bonds between atoms ['0', '5'] and with version 0"`
even though I can definitely see that there's an interaction there. I've tried various versions of the atoms argument in remove_interaction (give as an array, a list, a list of strings, etc.) but none of these seem to work.

I guess it might want me to give a version explicitly, but this doesn't seem to work.

Perhaps this should be raised as a separate issue, but on this last point when I read in the itp file that's been generated with martinize2, the meta information isn't read back in by vermouth - I think this might be quite a useful feature to have for this sort of purpose?

**Comment by pckroon** [comment]
Hey, 

I notice that the atoms you provide are a list of strings (`['0', '5']`). Did you try with a list of ints? 
Otherwise, take a look at what `ff.blocks[molname].interactions['bonds']` looks like.  Alternatively, it might be easier to iterate over all bonds, and filter them that way. 
Also, depending on how often you'll need to do something like this it might be easier for you to adapt/add the selection criterion you need to the elastic network processor (processors/apply_rubber_bands.py; take a look at how `make_same_region_criterion` is used for inspiration). If you think this functionality could be valuable to others as well (it does sound that way) we can also look at integrating this.

> Perhaps this should be raised as a separate issue, but on this last point when I read in the itp file that's been generated with martinize2, the meta information isn't read back in by vermouth - I think this might be quite a useful feature to have for this sort of purpose?

Yes please. It's definitely an omission on our part. It's been a while since I looked in detail, but renaming your itp to ff might also magically fix things. Or break things. But the file formats are very similar on purpose.

**Reply by csbrasnett** [reply]
Thanks Peter. Yes, I've tried a list of ints, etc, but I can't see anything that works this way. I'll have a go at looping over the bonds like you say. 

I'll open the issue from this discussion. I tried renaming the itp from molecule_0.itp to molecule_0.ff, but no change sadly.

**Reply by csbrasnett** [reply]
I've tried looping over the bonds in this way now, but I'm still getting the same KeyError as before.

```python
for i in ordered_domains:
    for j in ff.blocks[molname].interactions['bonds']:
        if j.atoms[0] > i[0] and j.atoms[0] < i[1] and j.atoms[1] > i[0] and j.atoms[1] < i[1]:
            continue
        else:
            ff.blocks[molname].remove_interaction('bonds', [j.atoms[0], j.atoms[1]])

````

**Reply by pckroon** [reply]
`            ff.blocks[molname].remove_interaction('bonds', j.atoms, version=j.meta.get('version', 0))` maybe? My guess now is that j.atoms is a tuple, and the tuple (a, b) != list [a, b]. It *is* rather annoying that `remove_interaction` is so picky.

As a (slightly pedantic) sidenote, I would recommend changing the variable names i and j. You'll thank yourself later. For example:
```py
for domain in ordered_domains:
    lower, upper = domain
    for bond in ff.blocks[molname].interactions['bonds']:
        atom1, atom2 = bond.atoms
        if lower <= atom1 < upper and lower <= atom2 < upper:  # if all(lower <= atom < upper for atom in bond.atoms):
            continue
        else:
            ff.blocks[molname].remove_interaction('bonds', bond.atoms, version=bond.meta.get('version', 0))
```

**Reply by csbrasnett** [reply]
That doesn't seem to work annoyingly! I checked, and j.atoms is actually a list, not a tuple. Either way, I tried casting both and neither worked.

ps - I was wrong above about reading in the meta sections. The bonds with ifdef FLEXIBLE attributes still have them when read in, but the distinct sections written into the itp file (side chain bonds, Backbone bonds, etc) aren't picked up.

**Reply by pckroon** [reply]
Hmmn. Could you print (part of) `ff.blocks[molname].interactions['bonds']` and post it here?

**Reply by csbrasnett** [reply]
Sure thing. I don't actually need to loop over the full dictionary entry, just the selection I manually find involved in the rubber band formation. Regardless, it looks like this:

 Interaction(atoms=[576, 629], parameters=['1', '0.76515', '700.0'], meta={}),
 Interaction(atoms=[576, 633], parameters=['1', '0.76011', '700.0'], meta={}),
 Interaction(atoms=[576, 635], parameters=['1', '0.69083', '700.0'], meta={}),
 Interaction(atoms=[576, 641], parameters=['1', '0.80783', '700.0'], meta={}),

**Reply by pckroon** [reply]
Ok, that looks sane at least. And if you now try `ff.blocks[molname].remove_interaction('bonds', [576, 629])`?
If *that* doesn't work we found a bug. If that does work something else is fishy

**Reply by csbrasnett** [reply]
No dice unfortunately. I can send you the script and pdb I'm working with if you want to take a look?

**Reply by pckroon** [reply]
Yes please. I'll have to see about freeing some time to have a look though

**Reply by csbrasnett** [reply]
Thanks Peter, apologies, I don't seem to have your email address?
